pkg/alerting/receiver.go

```diff
@@ -0,0 +1,202 @@
+	f.DurationVar(&cfg.PurgeInterval, "receiver.purge-interval", 15*time.Minute, "How often should we purge the in-memory measured timestamps tracker.")
+	f.DurationVar(&cfg.PurgeLookback, "receiver.purge-lookback", 2*time.Hour, "Period at which measured timestamps will remain in-memory.")
+}
+
+// NewReceiver returns an alertmanager webhook receiver
+func NewReceiver(cfg ReceiverConfig, log log.Logger, reg prometheus.Registerer) (*Receiver, error) {
+	lbl := make(map[string]string, 1)
+	if cfg.RoundtripLabel != "" {
+		l := strings.Split(cfg.RoundtripLabel, "=")
+
+		if len(l) != 2 {
+			return nil, fmt.Errorf("the label is not valid, it must have exactly one name and one value: %s has %d parts", l, len(l))
+		}
+
+		lbl[l[0]] = l[1]
+	}
+
+	r := &Receiver{
+		logger:     log,
+		cfg:        cfg,
+		timestamps: map[float64]struct{}{},
+		registry:   reg,
+	}
+
+	r.evalTotal = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Namespace: namespace,
+		Subsystem: "webhook_receiver",
+		Name:      "evaluations_total",
+		Help:      "The total number of evaluations made by the webhook receiver.",
+	})
+
+	r.failedEvalTotal = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Namespace: namespace,
+		Subsystem: "webhook_receiver",
+		Name:      "failed_evaluations_total",
+		Help:      "Total number of failed evaluations made by the webhook receiver.",
+	})
+
+	r.roundtripDuration = promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
+		Namespace:   namespace,
+		Subsystem:   "webhook_receiver",
+		Name:        "end_to_end_duration_seconds",
+		Help:        "Time spent (in seconds) from scraping a metric to receiving an alert.",
+		Buckets:     []float64{5, 15, 30, 60, 90, 120, 240},
+		ConstLabels: lbl,
+	}, []string{"alertname"})
+
+	r.wg.Add(1)
+	go r.purgeTimestamps()
+
+	return r, nil
+}
+
+// RegisterRoutes registers the receiver API routes with the provided router.
+func (r *Receiver) RegisterRoutes(router *mux.Router) {
+	router.Path("/api/v1/receiver").Methods(http.MethodPost).Handler(http.HandlerFunc(r.measureLatency))
+}
+
+func (r *Receiver) measureLatency(w http.ResponseWriter, req *http.Request) {
+	data := template.Data{}
+
+	// We want to take a snapshot of time the moment we receive the webhook
+	now := time.Now()
+
+	if err := json.NewDecoder(req.Body).Decode(&data); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		level.Error(r.logger).Log("msg", "unable to parse alerts", "err", err)
+		r.failedEvalTotal.Inc()
+		return
+	}
+
+	// We only care about firing alerts as part of this analysis.
+	for _, alert := range data.Alerts.Firing() {
+		var name string
+		for k, v := range alert.Labels {
+			if k == model.AlertNameLabel {
+				name = v
+			}
+		}
+
+		if name == "" {
+			level.Error(r.logger).Log("err", "alerts does not have an alertname label - we can't measure it", "labels", alert.Labels.Names())
+			continue
+		}
+
+		var t float64
+		for k, v := range alert.Annotations {
+			if k == "time" {
```

Comment:
This section is confusing. Assume we have 2 annotations, first one is `time` and second is `description`. According to the logic, we will first set it to `timestamp` in the first iteration, but then execute the logic again in the second interation. Maybe you meant something like:

```
for k, v := range alert.Annotations {
   if k != "time" {
     continue
   }

   // If its time, then execute logic.
}

// Maybe a check if we had no time in the annotations.

```